Add a note on build-time dependencies to INSTALL.md.

Zimmi48 opened the pull request:
Closes #12232.

cpitclaudel commented on INSTALL.md:
```diff
@@ -77,3 +77,14 @@ Please see [INSTALL.make.md](dev/doc/INSTALL.make.md) for build and
+- When loading plugins, you should make sure that these were compiled
+  with the same version of OCaml as Coq.  This is also true when
+  loading `vo` files.  For this reason, it is rarely a good idea to
+  distribute pre-compiled `vo` files.
```
It would be great to point out a way to determine which version of OCaml was used to compile a given .vo file.

Zimmi48 replied:
cc @ppedrot 

I don't think that's possible in fact. Even votour from Coq 8.11.1 fails on a vo file compiled with current master (and the converse is also true).

ppedrot replied:
We could store that information in the vo metadata. Anyways, there was recently a path by @ejgallego that stored the hash of the OCaml plugin files inside the vo, so that currently trying to load a vo with plugin dependencies mismatch would already trigger an error.

ppedrot replied:
See #12244.

SkySkimmer replied:
>This is also true when loading `vo` files

Is it really?

SkySkimmer replied:
>currently trying to load a vo with plugin dependencies mismatch would already trigger an error

That's not true, the hash is stored but not checked IIRC

andres-erbsen commented:
Thanks.

How about elaborating further on the limitation on loading plugins, perhaps `When loading plugins, you should make sure that these were compiled
  with the same version of OCaml as Coq. Note that plugin-Coq version mismatch can occur even if the plugin is successfully built right before attempting to load it -- having the target version of Coq installed while building is not enough.`?

Any pre-built distribution of coq likely includes the standard library .vo files. Is this really discouraged?

ejgallego commented:
As written, this is confusing and actually not even accurate.

We should direct our energy to fixing broken build systems instead of trying to add some weird partial explanations that will confuse people more than help.

andres-erbsen commented:
I agree that the "fixes" here are likely going to involve users changing their "broken" workflows, but I think that providing an accurate specification of what is required is a necessary prerequisite for that. Perhaps what needs to be said is indeed something that "every ocaml user should know", but every coq user is not a (knowing) ocaml user, even if this user uses plugins. If we can link to ocaml documentation that explains the situation instead of writing our own, great. I don't think that ignoring the issue entirely is going to do anyone a favor, and the proposed text would definitely have saved me a good bit of time in the past.

Zimmi48 commented:
> Any pre-built distribution of coq likely includes the standard library .vo files. Is this really discouraged?

Indeed, it is not. What is discouraged is distributing `.vo` files alone.

Zimmi48 commented:
> As written, this is confusing and actually not even accurate.

Please help me improve it! The fact that even I cannot write a fully accurate description of the situation shows how little this knowledge is shared and thus how important it is to get something written down.

> We should direct our energy to fixing broken build systems instead of trying to add some weird partial explanations that will confuse people more than help.

Who is going to put energy in fixing broken build systems? You said yourself "My build system / setup of choice properly tracks dependencies so I have not seen this error since a long time" in https://github.com/coq/coq/pull/12244#issuecomment-623194021, so it doesn't seem likely that you will be the one spending time to fix *other* build systems. If you want other people to be able to do that, or to avoid any build system besides the ones that are vetted by the development team, then something should definitely be written.

ejgallego commented:
> What is discouraged is distributing `.vo` files alone.

Why is this discouraged, and where? There is no problem in distributing the `.vo` files if they have been built properly.

Zimmi48 commented:
> Why is this discouraged, and where? There is no problem in distributing the `.vo` files if they have been built properly.

But you should document precisely with which version of Coq compiled with which version of OCaml you should load them, correct?

ejgallego commented:
> Please help me improve it! The fact that even I cannot write a fully accurate description of the situation shows how little this knowledge is shared and thus how important it is to get something written down.

Well, that's a bit of my point; this is OCaml-specific stuff so I think we should not DIY and instead rely on upstream tools.

The original error did happen if you roll your own build / package system.

ejgallego commented:
> But you should document precisely with which version of Coq compiled with which version of OCaml you should load them, correct?

They have to be built with the same exact toolchain, but that doesn't impede [as we discussed] for example library authors setting a CI job that uses the docker image used to build Coq to generate the proper binaries; that's for example how jsCoq alpha package build system works and how we plan to do for windows / OSX too.

ejgallego commented:
Don't take my concerns as blocking this PR, if people feel there is an improvement OK, tho I still feel we will make things more confusing. True rules are a tad complex tho, for example even if you use different OCaml options you may need to rebuild everything.

ejgallego commented on INSTALL.md:
```diff
@@ -77,3 +77,14 @@ Please see [INSTALL.make.md](dev/doc/INSTALL.make.md) for build and
+  run-time dependencies if you wish to use the native compiler.
+
+- When loading plugins, you should make sure that these were compiled
+  with the same version of OCaml as Coq.  This is also true when
```
I'd write "same OCaml setup" instead of version. In particular, all libraries and transitive dependencies have to match.

cpitclaudel commented:
I think documenting this is very useful.

> They have to be built with the same exact toolchain, but that doesn't impede

Doesn't it though? At work we use a shared machine to pair-program remotely on Coq code.  The shared machine isn't particularly beefy, so building there isn't great.  For now I just rsync vo files onto it, and knowing when that is supposed to work is good.

JasonGross commented:
> > They have to be built with the same exact toolchain, but that doesn't impede
> 
> Doesn't it though? At work we use a shared machine to pair-program remotely on Coq code. The shared machine isn't particularly beefy, so building there isn't great. For now I just rsync vo files onto it, and knowing when that is supposed to work is good.

If both machines are running the same version of Ubuntu/Linux, are both using opam-installed OCaml, and are both using the same version of OCaml, and the same version of Coq which was built by that version of OCaml, then it should be fine.  I am not sure about the case of using different versions of Ubuntu/Linux.  I am quite sure that using different versions of Coq will break things, and I'm pretty sure that using different versions of OCaml will break things.  I am not sure if it will break things if one machine has a hand-compiled OCaml and the other an opam-compiled OCaml; I suspect if you get the flags exactly right, it'll work, and otherwise it might not.

Zimmi48 commented:
> The original error did happen if you roll your own build / package system.

I don't know what you mean by "roll your own package system" but are you saying that people should not package Coq in Linux distributions and only ever use opam?! o_O

> I'd write "same OCaml setup" instead of version. In particular, all libraries and transitive dependencies have to match, as well as usually options and environment variables affecting the compiler.

OK, I can change this.

Zimmi48 commented:
I've also made the claim about vo files more specific.

ejgallego commented:
> are you saying that people should not package Coq in Linux distributions and only ever use opam?

No. I am saying that if you roll your own package system it is also your responsibility to ensure that its build rules are correct.

Zimmi48 commented:
My point was just that there is a wide variety of package systems and that saying "roll your own" is a bit dismissive of this reality. People packaging Coq for these (existing) package systems might not be OCaml expert nor aware of the limitations. So it is important to make them aware (and `INSTALL.md` is probably the file they will look into for knowledge). Of course, we should link to upstream documentation every time we can (especially if such documentation does exist).

ejgallego commented:
> People packaging Coq for these (existing) package systems might not be OCaml experts nor aware of the limitations. So it is important to make them aware (and `INSTALL.md` is probably the file they will look into for knowledge). Of course, we should link to upstream documentation every time we can (especially if such documentation does exist).

I don't follow this logic, people packaging Coq should understand how OCaml packaging work, otherwise I don't see how they could produce a working package.

My point is that this kind of documentation doesn't belong in Coq; but in OCaml, imagine every C application having to explain how C ABI and linking works...

Zimmi48 commented:
People packaging software for Linux distributions are usually just volunteers doing their best.

The Arch case described in ocaml/ocaml#9532 is pretty interesting I think: OCaml was updated without Coq being rebuilt. Coq still works perfectly fine (so this goes unnoticed) except when you try to build a plugin. If you have a system-installed OCaml, suddenly you cannot build a plugin and have it work with your system-installed Coq!

Maybe the note should be even more specific to cover this case.

@andres-erbsen Does native compute run fine or fails horribly in your system-installed Coq on Arch?

andres-erbsen commented:
Nothing fails horribly, but I guess I wouldn't know if it wasn't actually using the native compiler.

```
browse@ashryn ~ % cat /tmp/a.v
Eval native_compute in 1+1.
browse@ashryn ~ % coqc /tmp/a.v                                  
     = 2     : nat
```

ejgallego commented:
> People packaging software for Linux distributions are usually just volunteers doing their best.

Who said otherwise?

This still doesn't make my point less true, people packaging OCaml programs should refer to OCaml documentation not to Coq documentation.

I am OK if you want to add a note but in the end it is gonna solve little for packagers and they will have to refer to the OCaml documentation anyways; in the worst case the note in Coq will be misguided / out of date.

I am not familiar with Arch, Debian used to have to encode the hash of OCaml libs in the package version so indeed it seems OCaml is a tricky language to build / package ; I am not so expert myself.

JasonGross commented on INSTALL.md:
```diff
@@ -77,3 +77,16 @@ Please see [INSTALL.make.md](dev/doc/INSTALL.make.md) for build and
+  with the same OCaml setup (version, flags, dependencies...) as Coq.
+  This is also true when loading `vo` files.  For this reason, it is
+  rarely a good idea to distribute pre-compiled `vo` files to the
+  general public (if you don't know what software they will use to
+  load them).
```
This makes it sound like a bit like they might be using something other than Coq to load vo files.  I'd suggest
```suggestion
  general public (if you don't know whether they'll have the same
  OCaml and Coq setup as was used for compilation).
```

Zimmi48 commented:
> This still doesn't make my point less true, people packaging OCaml programs should refer to OCaml documentation not to Coq documentation.

Sure, but Coq is a special bird. Not all OCaml applications have plugins. And I would be happy to point to the relevant OCaml documentation instead of writing approximate things *if I knew where it is*.

ejgallego commented on INSTALL.md:
```diff
@@ -77,3 +77,16 @@ Please see [INSTALL.make.md](dev/doc/INSTALL.make.md) for build and
 the Dune-based system see the [dune guide for
 developers](dev/doc/build-system.dune.md).
+
+Notes on build-time dependencies
```
Not sure "build-time dependencies" is the right wording here; already above we discuss about what we need to build, and just below we talk about run-time dependencies.

I suggest:
- we move the native dependencies to its own section, "Dependencies of native compilation"
- we remove the second bullet, and make it into a section "OCaml toolchain advisory" , also I would just start with "Both Coq plugins and vo files require at loading time the same exact toolchain setup that was used to compile them, etc..."

Other than that, I'd like to merge this before the branch date.

Zimmi48 commented:
Updated following your advice. I've removed the part saying:

>For this reason, it is rarely a good idea to distribute pre-compiled vo files to the public (if you don't know whether they'll have the same OCaml and Coq setup as was used for compilation).

because I was under the impression that you felt it is too much, or redundant with the preceding explanation. But I can reintroduce it if that was a mistake.

ejgallego commented on INSTALL.md:
```diff
@@ -77,3 +77,16 @@ Please see [INSTALL.make.md](dev/doc/INSTALL.make.md) for build and
+
+When loading plugins or `vo` files, you should make sure that these
+were compiled with the same OCaml setup (version, flags,
+dependencies...) as Coq.
```
It just occurred to me a couple of further comments [sorry for the churn]
- we don't mention the actual error anywhere it seems. So users may have trouble mapping the dreaded "implementation mismatch on Foo" to this error.
- should something be also added to reference manual?

Regarding your comment on distribution, I dunno. Maybe I would phrase it positively "distribution of pre-compiled plugins and `.vo` files is only possible if users are guaranteed to have the exact same OCaml toolchain".

ejgallego commented:
Also note that technically this doesn't seem to close #12232 as that bug talks about the refman.

ejgallego commented:
Maybe it is time we move `INSTALL` to the refman?

Zimmi48 commented:
> Also note that technically this doesn't seem to close #12232 as that bug talks about the refman.

Yes, but I disagreed about including this info in the refman.

> Maybe it is time we move `INSTALL` to the refman?

To me this is a legitimate exception because this documentation must be readable from the sources, since it contains instructions to build Coq (and building the refman, which is documented in `doc/README.md`, requires building Coq).

>we don't mention the actual error anywhere it seems. So users may have trouble mapping the dreaded "implementation mismatch on Foo" to this error.

Right, but this is an OCaml error message, which doesn't seem legitimate to add to the Coq documentation (as it could change, etc.).

> Maybe I would phrase it positively "distribution of pre-compiled plugins and `.vo` files is only possible if users are guaranteed to have the exact same OCaml and Coq toolchain".

Sounds good. I'll add this sentence.

ejgallego commented:
Ok to all the points @Zimmi48 , except:

> Right, but this is an OCaml error message, which doesn't seem legitimate to add to the Coq documentation (as it could change, etc.).

I think the message has been stable for more than a decade, so I still suggest we add it. IMHO users may have otherwise a hard time finding this info in `INSTALL` file, as one of my gripes with this message is that the problem will rarely happen at install time, on the contrary, you tend to encounter this message long after you have installed Coq, due to bugs in package/build systems.

Zimmi48 commented:
Updated with the error message.
